scipy/sparse/csgraph/_flow.pyx

```diff
@@ -670,3 +685,905 @@ cdef ITYPE_t[:] _dinic(
+    for v in range(1, n_verts + 1):
+        parent[v] = 0
+        parent_edge[v] = n_non_zero_edges + v - 1
+        subtree_size[v] = 1
+        prev_vertex_dft[v] = v - 1
+        last_descendent_dft[v] = v
+
+    # next_vertex_dft initialization
+    for v in range(n_verts):
+        next_vertex_dft[v] = v + 1
+    next_vertex_dft[n_verts] = 0
+
+
+cdef inline ITYPE_t _reduced_cost(
+    ITYPE_t e,                       # IN
+    ITYPE_t[:] edge_weights,         # IN
+    ITYPE_t[:] vertex_potentials,    # IN
+    ITYPE_t[:] edge_sources,         # IN
+    ITYPE_t[:] edge_targets,         # IN
+    ITYPE_t[:] edge_flow             # IN
+) nogil:
+    """Returns the reduced cost of an edge computed using
+    potentials of the vertices of the given edge and its weight.
+
+    The reduced cost is needed to determine whether adding the
+    edge to current spanning tree improves circulation or not.
+    """
+    cdef ITYPE_t c = (edge_weights[e]
+         - vertex_potentials[edge_sources[e]]
+         + vertex_potentials[edge_targets[e]]
+    )
+    return c if edge_flow[e] == 0 else -c
+
+ctypedef struct local_vars:
+    # A structure to store local
+    # variables used in _find_entering_edges
+    # function.
+    # This structure is needed to implement
+    # the above function without using yield
+    # feature of Python.
+    ITYPE_t B
+    ITYPE_t M
+    ITYPE_t m
+    ITYPE_t f
+
+cdef class network_simplex_result:
+    """
+    Cython class to store the network
+    simplex result.
+    """
+    cdef readonly ITYPE_t flow_cost
+    cdef readonly ITYPE_t size
+    cdef readonly bint is_correct
+
+ctypedef struct edge_result:
+    # A structure to store edges found out
+    # by _find_entering_edges function.
+    # This structure is needed to implement
+    # the above function without using yield
+    # feature of Python.
+    ITYPE_t i
+    ITYPE_t p
+    ITYPE_t q
+
+ctypedef struct return_struct:
+    # A structure to store the return value
+    # and current state of _find_entering_edges
+    # function.
+    # This structure is needed to implement
+    # the above function without using yield
+    # feature of Python.
+    local_vars B_M_m_f
+    edge_result i_p_q
+    bint prev_ret_value
+
+cdef return_struct _find_entering_edges(
+    ITYPE_t n_edges,                     # IN
+    ITYPE_t[:] edge_weights,             # IN
+    ITYPE_t[:] vertex_potentials,        # IN
+    ITYPE_t[:] edge_flow,                # IN
+    ITYPE_t[:] edge_sources,             # IN
+    ITYPE_t[:] edge_targets,             # IN
+    return_struct r_struct,              # IN
+) nogil:
+    """Pivoting strategy of the network simplex algorithm.
+
+    We have used the NetworkX implementation of this function.
+    However, we don't use yield and instead manually maintain
+    the state of this function using return_struct defined above.
+
+    Following is the explanation from `networkx.algorithms.flow.networksimplex`:
```

Comment:
minor: `networkx` is not linked via intersphinx, so this won't resolve if it's every included in html docs - please use double backticks to typeset as code.